crates/ibc/src/core/ics04_channel/handler/acknowledgement.rs

```diff
@@ -11,6 +11,130 @@ use crate::events::IbcEvent;
+            }
+            .into());
+        }
+
+        // Verify packet commitment
+        let commitment_on_a = match ctx_a.get_packet_commitment(&(
+            msg.packet.port_on_a.clone(),
+            msg.packet.chan_on_a.clone(),
+            msg.packet.sequence,
+        )) {
+            Ok(commitment_on_a) => commitment_on_a,
+
+            // This error indicates that the timeout has already been relayed
+            // or there is a misconfigured relayer attempting to prove a timeout
+            // for a packet never sent. Core IBC will treat this error as a no-op in order to
+            // prevent an entire relay transaction from failing and consuming unnecessary fees.
+            Err(_) => return Ok(()),
+        };
+
+        if commitment_on_a
+            != ctx_a.packet_commitment(
+                &packet.data,
+                &packet.timeout_height_on_b,
+                &packet.timeout_timestamp_on_b,
+            )
+        {
+            return Err(PacketError::IncorrectPacketCommitment {
+                sequence: packet.sequence,
+            }
+            .into());
+        }
+
+        // Verify proofs
+        {
+            let client_id_on_a = conn_end_on_a.client_id();
+            let client_state_on_a = ctx_a.client_state(client_id_on_a)?;
+
+            // The client must not be frozen.
+            if client_state_on_a.is_frozen() {
+                return Err(PacketError::FrozenClient {
+                    client_id: client_id_on_a.clone(),
+                }
+                .into());
+            }
+
+            let consensus_state = ctx_a.consensus_state(client_id_on_a, &msg.proof_height_on_b)?;
+
+            let ack_commitment = ctx_a.ack_commitment(&msg.acknowledgement);
+
+            // Verify the proof for the packet against the chain store.
+            client_state_on_a
+                .new_verify_packet_acknowledgement(
+                    ctx_a,
+                    msg.proof_height_on_b,
+                    &conn_end_on_a,
+                    &msg.proof_acked_on_b,
+                    consensus_state.root(),
+                    &packet.port_on_b,
+                    &packet.chan_on_b,
+                    packet.sequence,
+                    ack_commitment,
+                )
+                .map_err(|e| ChannelError::PacketVerificationFailed {
+                    sequence: packet.sequence,
+                    client_error: e,
+                })
+                .map_err(PacketError::Channel)?;
+        }
+
+        if let Order::Ordered = chan_end_on_a.ordering {
+            let next_seq_ack = ctx_a.get_next_sequence_ack(port_chan_id_on_a)?;
+
+            if packet.sequence != next_seq_ack {
+                return Err(PacketError::InvalidPacketSequence {
+                    given_sequence: packet.sequence,
+                    next_sequence: next_seq_ack,
+                }
+                .into());
+            }
+        }
```

Comment:
Wouldn't be better to move this check before the "Verify proofs" section so that it stops in error cases before doing that more resource-intensive proof verification step?